cli/fathom_web/commands/test.py

```diff
@@ -0,0 +1,77 @@
+from json import JSONDecodeError, load, loads
+
+from click import argument, BadParameter, command, File, style
+
+from ..accuracy import accuracy_per_tag, accuracy_per_page, pretty_accuracy
+from ..utils import classifier, tensor, tensors_from
+
+
+def decode_weights(ctx, param, value):
+    """Validate a click option, making sure it's a valid JSON object with
+    properly formatted "coeff" and "bias" keys."""
+    try:
+        decoded_weights = loads(value)
+    except JSONDecodeError:
+        raise BadParameter('Weights must be a valid JSON object.')
+    else:
+        if 'coeffs' in decoded_weights and 'bias' in decoded_weights:
+            if type(decoded_weights['bias']) is float:
+                if (type(decoded_weights['coeffs']) is list and
+                    all((len(pair) == 2 and type(pair[0]) is str and type(pair[1]) is float)
```

Comment:
This if-clause is a bit unwieldy. Pulling the whole thing out into its own method would make it easier to read at the expense of the additional lines. I think this is personal preference.
```
def coeffs_not_list_of_2_element_lists(coeffs):
    if not isinstance(coeffs, list):
        return False
    for pair in coeffs:
        if len(pair) != 2:
            return False
        if not isinstance(pair[0], str):
            return False
        if not isinstance(pair[1], float):
            return False
    return True

...

if coeffs_not_list_of_2_element_lists(decoded_weights['coeffs']):
    raise BadParameter('Coeffs must be a list of 2-element lists: [["ruleName", numericCoefficient], ...].')
```

Reply:
I rewrapped the predicates in the generator expression. Hopefully that makes it clearer?